File: spring/src/main/java/com/ciel/domain/Post.java
package com.ciel.domain;

import com.fasterxml.jackson.annotation.JsonInclude;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.OffsetDateTime;
import java.util.List;
import java.util.UUID;

/** Mirrors Rust's {@code domain::post::Post}. */
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@JsonInclude(JsonInclude.Include.ALWAYS)
public class Post {
    private UUID id;
    private UUID ownerId;
    private String ownerHandle;
    private String ownerDisplayName;
    private List<UUID> mediaIds;

    @JsonInclude(JsonInclude.Include.NON_NULL)
    private Media primaryMedia;

    private String caption;
    private OffsetDateTime createdAt;
    private PostVisibility visibility;

    @JsonInclude(JsonInclude.Include.NON_NULL)
    private String ownerAvatarKey;

    @JsonInclude(JsonInclude.Include.NON_NULL)
    private String ownerAvatarUrl;

    @JsonInclude(JsonInclude.Include.NON_NULL)
    private Long likeCount;

    @JsonInclude(JsonInclude.Include.NON_NULL)
    private Long commentCount;

    @JsonInclude(JsonInclude.Include.NON_NULL)
    private Boolean likedByViewer;
}

File: dotnet/Ciel.Api/Domain/Post.cs
using System.Text.Json.Serialization;

namespace Ciel.Api.Domain;

public sealed class Post
{
    public Guid Id { get; set; }
    public Guid OwnerId { get; set; }
    public string? OwnerHandle { get; set; }
    public string? OwnerDisplayName { get; set; }
    public List<Guid> MediaIds { get; set; } = [];

    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public Media? PrimaryMedia { get; set; }

    public string? Caption { get; set; }
    public DateTimeOffset CreatedAt { get; set; }
    public PostVisibility Visibility { get; set; }

    // Rust's `Post::owner_avatar_key` uses `skip_serializing_if = "Option::is_none"`,
    // so it IS included on the wire (and in the Redis feed cache) whenever present.
    // A previous unconditional [JsonIgnore] here dropped it from the cache entirely,
    // breaking avatar URL population on cache hits.
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? OwnerAvatarKey { get; set; }

    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? OwnerAvatarUrl { get; set; }

    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public long? LikeCount { get; set; }

    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public long? CommentCount { get; set; }

    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public bool? LikedByViewer { get; set; }
}
